pkg/istiod/common.go

```diff
@@ -0,0 +1,324 @@
+	// for testing.
+	basePort           int32
+	secureGrpcListener net.Listener
+}
+
+func (s *Server) InitCommon(args *PilotArgs) {
+
+	_, addr, err := startMonitor(args.DiscoveryOptions.MonitoringAddr, s.mux)
+	if err != nil {
+		return
+	}
+	s.MonitorListeningAddr = addr
+
+	//go func() {
+	//	<-s.stop
+	//	err := monitor.Close()
+	//	log.Debugf("Monitoring server terminated: %v", err)
+	//}()
+
+}
+
+// Start all components of istio, using local config files or defaults.
+//
+// A minimal set of Istio Env variables are also used.
+// This is expected to run in a Docker or K8S environment, with a volume with user configs mounted.
+//
+// Defaults:
+// - http port 15007
+// - grpc on 15010
+//- config from $ISTIO_CONFIG or ./conf
+func InitConfig(confDir string) (*Server, error) {
+	baseDir := "." // TODO: env ISTIO_HOME or HOME ?
+
+	// TODO: 15006 can't be configured currently
+	// TODO: 15090 (prometheus) can't be configured. It's in the bootstrap file, so easy to replace
+
+	meshCfgFile :=  baseDir + confDir + "/mesh"
+
+	// Create a test pilot discovery service configured to watch the tempDir.
+	args := &PilotArgs {
+		DomainSuffix: "cluster.local",
+
+		Mesh: MeshArgs{
+			ConfigFile: meshCfgFile,
+			RdsRefreshDelay: types.DurationProto(10 * time.Millisecond),
+		},
+		Config: ConfigArgs{},
+
+		// MCP is messing up with the grpc settings...
+		MCPMaxMessageSize:        1024 * 1024 * 64,
+		MCPInitialWindowSize:     1024 * 1024 * 64,
+		MCPInitialConnWindowSize: 1024 * 1024 * 64,
+	}
+
+	// Main server - pilot, registries
+	server, err := NewServer(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := server.WatchMeshConfig(meshCfgFile); err != nil {
+		return nil, fmt.Errorf("mesh: %v", err)
+	}
+
+	pilotAddress := server.Mesh.DefaultConfig.DiscoveryAddress
+	_, port, _ := net.SplitHostPort(pilotAddress)
+	basePortI, _ := strconv.Atoi(port)
+	basePortI = basePortI - basePortI % 100
+	basePort := int32(basePortI)
+	server.basePort = basePort
+
+
+	args.DiscoveryOptions =  envoy.DiscoveryServiceOptions{
+		HTTPAddr:        fmt.Sprintf(":%d", basePort+7),
+		GrpcAddr:        fmt.Sprintf(":%d", basePort+10),
+		// Using 12 for K8S-DNS based cert.
+		// TODO: We'll also need 11 for Citadel-based cert
+		SecureGrpcAddr:        fmt.Sprintf(":%d", basePort+12),
+		EnableCaching:   true,
+		EnableProfiling: true,
+	}
+	args.CtrlZOptions = &ctrlz.Options{
+		Address: "localhost",
+		Port:    uint16(basePort + 13),
+	}
+
+		err = server.InitConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Galley args
+	gargs := settings.DefaultArgs()
+
+	// Default dir.
+	// If not set, will attempt to use K8S.
+	gargs.ConfigPath = baseDir + "/var/lib/istio/local"
+	// TODO: load a json file to override defaults (for all components)
+
+	gargs.ValidationArgs.EnableValidation = false
+	gargs.ValidationArgs.EnableReconcileWebhookConfiguration = false
+	gargs.APIAddress = fmt.Sprintf("tcp://0.0.0.0:%d", basePort+901)
```

Comment:
all these random numbers are going to cause confusion, should at least have a comment somewhere what all the ports are maybe?

Reply:
I agree. Might be more clear if we separate out the base port code into a separate structure,  which has all the ports defined.